File: js/feedback.js
/* Vibration and short synthesised blips, both gated on settings and throttled
   so a bullet-hell frame can't fire hundreds of them. */
(function (global) {
  'use strict';

  var CUES = {
    purchase: { freq: 660, to: 990, dur: 0.09, type: 'triangle', gain: 0.16 },
    ability:  { freq: 220, to: 720, dur: 0.22, type: 'sawtooth', gain: 0.14 },
    travel:   { freq: 420, to: 880, dur: 0.30, type: 'sine', gain: 0.14 },
    denied:   { freq: 200, to: 130, dur: 0.14, type: 'square', gain: 0.09 },
    rebirth:  { freq: 320, to: 1200, dur: 0.45, type: 'sine', gain: 0.16 },
    pop:      { freq: 880, to: 1320, dur: 0.05, type: 'sine', gain: 0.07 }
  };

  var Feedback = {
    hapticsEnabled: true,
    soundEnabled: true,
    _ctx: null,
    _lastHaptic: 0,
    _lastSound: 0,
    /** Browsers reject vibration and audio until the page has been tapped. */
    _gestureSeen: false,

    /** Browsers only allow audio after a user gesture; call this from one. */
    unlock: function () {
      this._gestureSeen = true;
      if (!this._ctx) {
        var Ctx = global.AudioContext || global.webkitAudioContext;
        if (!Ctx) return;
        try { this._ctx = new Ctx(); } catch (e) { this._ctx = null; return; }
      }
      if (this._ctx.state === 'suspended') this._ctx.resume();
    },

    tap: function (strength, throttle) {
      if (!this.hapticsEnabled || !this._gestureSeen) return;
      if (!global.navigator || !navigator.vibrate) return;
      // Chrome logs a console error if we call this before a real gesture.
      if (navigator.userActivation && !navigator.userActivation.hasBeenActive) return;
      var now = performance.now();
      var gap = (throttle === undefined ? 0.05 : throttle) * 1000;
      if (now - this._lastHaptic < gap) return;
      this._lastHaptic = now;
      var ms = strength === 'heavy' ? 22 : (strength === 'medium' ? 12 : 7);
      try { navigator.vibrate(ms); } catch (e) { /* not supported */ }
    },

    select: function () { this.tap('light', 0.02); },
    success: function () { this.tap('medium', 0.02); },

    play: function (cueName, throttle) {
      if (!this.soundEnabled) return;
      var now = performance.now();
      var gap = (throttle === undefined ? 0.08 : throttle) * 1000;
      if (now - this._lastSound < gap) return;
      this._lastSound = now;

      this.unlock();
      var ctx = this._ctx;
      if (!ctx || ctx.state !== 'running') return;

      var cue = CUES[cueName];
      if (!cue) return;

      var osc = ctx.createOscillator();
      var gain = ctx.createGain();
      var t = ctx.currentTime;

      osc.type = cue.type;
      osc.frequency.setValueAtTime(cue.freq, t);
      osc.frequency.exponentialRampToValueAtTime(Math.max(20, cue.to), t + cue.dur);

      gain.gain.setValueAtTime(0.0001, t);
      gain.gain.exponentialRampToValueAtTime(cue.gain, t + 0.012);
      gain.gain.exponentialRampToValueAtTime(0.0001, t + cue.dur);

      osc.connect(gain);
      gain.connect(ctx.destination);
      osc.start(t);
      osc.stop(t + cue.dur + 0.02);
    }
  };

  global.Feedback = Feedback;
})(window);

File: IdleDotShooter/Services/Feedback.swift
import AudioToolbox
import QuartzCore
import UIKit

/// Haptics and UI sound, both gated on the player's settings and throttled so a
/// bullet-hell frame can't machine-gun the taptic engine.
final class Feedback {

    static let shared = Feedback()

    var hapticsEnabled = true
    var soundEnabled = true

    private let light = UIImpactFeedbackGenerator(style: .light)
    private let medium = UIImpactFeedbackGenerator(style: .medium)
    private let heavy = UIImpactFeedbackGenerator(style: .heavy)
    private let selection = UISelectionFeedbackGenerator()
    private let notice = UINotificationFeedbackGenerator()

    private var lastHaptic: TimeInterval = 0
    private var lastSound: TimeInterval = 0

    private init() {}

    func prepare() {
        light.prepare()
        medium.prepare()
        selection.prepare()
    }

    enum Strength { case light, medium, heavy }

    func tap(_ strength: Strength = .light, throttle: TimeInterval = 0.05) {
        guard hapticsEnabled else { return }
        let now = CACurrentMediaTime()
        guard now - lastHaptic >= throttle else { return }
        lastHaptic = now
        switch strength {
        case .light:  light.impactOccurred()
        case .medium: medium.impactOccurred()
        case .heavy:  heavy.impactOccurred()
        }
    }

    func select() {
        guard hapticsEnabled else { return }
        selection.selectionChanged()
    }

    func success() {
        guard hapticsEnabled else { return }
        notice.notificationOccurred(.success)
    }

    func warning() {
        guard hapticsEnabled else { return }
        notice.notificationOccurred(.warning)
    }

    enum Cue: SystemSoundID {
        case purchase = 1_104
        case ability = 1_113
        case travel = 1_057
        case denied = 1_053
        case rebirth = 1_025
    }

    func play(_ cue: Cue, throttle: TimeInterval = 0.08) {
        guard soundEnabled else { return }
        let now = CACurrentMediaTime()
        guard now - lastSound >= throttle else { return }
        lastSound = now
        AudioServicesPlaySystemSound(cue.rawValue)
    }
}
